### DISCUSSION TITLE
Use prisma.raw with .sql file generated by pg_dump

### DISCUSSION THREAD

**Discussion by BjoernRave** [prisma/prisma#2528] [Q&A] [answered]
Hey, I am setting up integration tests for my application and I would like to apply a `.sql` file after every test to have fresh data.

Does anyone have any success using `prisma.raw` to execute the `.sql`file ? 

I have this code so far and it also runs, but it doesn't actually insert anything.

```
  const sql = fs
    .readFileSync('./test.sql')
    .toString()
    .replace(/(\r\n|\n|\r)/gm, ' ') // remove newlines
    .replace(/\s+/g, ' ') // excess white space

  await prisma.raw(sql)
```

**Comment by Sytten** [comment]
The raw support is currently very limited, so I am curious as well if that is possible.
Can you enable debug mode (`DEBUG=*`) and take a look at the queries that are performed.

**Comment by ryands17** [comment]
@BjoernRave this approach works for me.

**Reply by BjoernRave** [reply]
Ah really? Are you using postgres? What command are you running exactly to generate the SQL file?

**Reply by ryands17** [reply]
Tested on SQLite and Postgres both with an insert statement as multiple commands are not supported directly. For that you would have to split by `\n` and create an array and then insert those array values.

**Reply by BjoernRave** [reply]
ah, that might have been the problem. Thanks for the pointer, I will try that

**Comment by BjoernRave** [comment]
okay, so I tried around a bit and I am using this command to do the backup now:
```
pg_dump --no-publications --no-security-labels --no-subscriptions -O --no-tablespaces --no-comments -a -d postgresql://bjoern@localhost:5432/dev > test.sql
```
Now my `.sql` file almost only consists of `COPY` statements. Splitting at `/n` did not work, since some commands are multiple lines.

This is the code I came up with:
```
  const sqls = fs
    .readFileSync('./test.sql')
    .toString()
    .split('\n')
    .filter((line) => line.indexOf('--') !== 0)
    .join('\n')
    .replace(/(\r\n|\n|\r)/gm, ' ') // remove newlines
    .replace(/\s+/g, ' ') // excess white space
    .split('\\.')

  for (const sql of sqls) {
    try {
      await prisma.raw(sql)
    } catch (error) {}
  }
```

This is a snippet of what my sql file looks like:
```

--
-- Data for Name: Address; Type: TABLE DATA; Schema: public; Owner: -
--

COPY public."Address" ("addressLine1", "addressLine2", city, country, id, state, zip) FROM stdin;
34369 Jordon Cape	\N	West Darylland	Equatorial Guinea	ckagkvsdo0001vtv4tsj79ka7	Ohio	53893
02433 Maximilian Mount	\N	South Arnaldostad	Faroe Islands	ckagkvsdv0003vtv4bswx9all	Indiana	48781-4146
964 Skylar Shore	\N	Port Jake	Panama	ckagkvse10005vtv4as9tfm1b	Pennsylvania	74678
45132 Bernice Viaduct	\N	Howellton	United Kingdom	ckagkvseb0007vtv4khr6e607	Minnesota	65544
5714 Annamarie Forest	\N	South Bransonland	Malawi	ckagkvseh0009vtv48s5zzoze	North Dakota	78284-7640
4891 Littel Springs	\N	Port Haylee	Turkey	ckagkvsem0011vtv4t94u1wxy	Connecticut	19066-8912
9426 Elvie Circle	\N	Port Naomi	Algeria	ckagkvsfa0022vtv4jsxjaipc	New York	50743-3324
77440 Klocko Parkways	\N	Harmontown	United Kingdom	ckagkvsfg0024vtv4uheihl2e	Maine	60001
\.


--
-- Data for Name: Import; Type: TABLE DATA; Schema: public; Owner: -
--

COPY public."Import" ("createdAt", id) FROM stdin;
\.
```
(No worries, this is just faker data)

So I am at first splitting at `\n` to get rid of all the comments. 

Then I am splitting at `\.` to get each statement.

If I only run the first copy block alone I get the following error: 
```
UnhandledPromiseRejectionWarning: Error: Raw query failed. Code: `42601`. Message: `db error: ERROR: syntax error at or near "34369"`
```

I guess the problem comes from the "stdin" part, but not very experienced with that

**Comment by BjoernRave** [comment, marked as answer]
Okay I looked up what the `stdin` part actually means and it turns out it's a psql specific thing.

So I added `--column-inserts --attribute-inserts` to my `pg_dump` turning all `COPY` into explicit `INSERT INTO` statements and now it's working 🥳

I think those two and the `-a` flag, which makes it only dump the data are the most important ones

So my final dump command is: 
```
pg_dump --no-publications --no-security-labels --no-subscriptions -O --no-tablespaces --no-comments -a --column-inserts --attribute-inserts -d postgresql://bjoern@localhost:5432/dev > test.sql
```

(Not sure if I actually need all those `--no-*` flags, just wanted to be sure :D)

and my code is: 
```
  const sqls = fs
    .readFileSync('./test.sql')
    .toString()
    .split('\n')
    .filter((line) => line.indexOf('--') !== 0)
    .join('\n')
    .replace(/(\r\n|\n|\r)/gm, ' ') // remove newlines
    .replace(/\s+/g, ' ') // excess white space
    .split(';')

  for (const sql of sqls) {
    await prisma.raw(sql)
  }
```

**Reply by patricktree** [reply]
Thank you for this script!
It works also with Prisma 3.10.0 but using `prisma.$executeRawUnsafe`.

I had one issue tough: my SQL script contains strings, which themselves sometimes contain a semicolon.  
Like so:

```sql
INSERT INTO public."KnowHowQuestion" (id, topic) VALUES (30, 'Börse; Aktien');
```

Your suggested code uses `.split(';')`, this would split that string apart, thus such statements were no valid SQL anymore.

I implemented a function which splits at `;` only if it is not inside of a string.  
I also applied some ESLint suggestions, my final `seed.ts` script is this:

```typescript
/* eslint-disable no-console, no-process-exit */
import fs from 'fs';
import path from 'path';

import { PrismaClient } from '~/lib/prisma-client';

const prisma = new PrismaClient();

async function main() {
  try {
    const rawSql = await fs.promises.readFile(path.join(__dirname, '../../sql/seed.sql'), {
      encoding: 'utf-8',
    });
    const sqlReducedToStatements = rawSql
      .split('\n')
      .filter((line) => !line.startsWith('--')) // remove comments-only lines
      .join('\n')
      .replace(/\r\n|\n|\r/g, ' ') // remove newlines
      .replace(/\s+/g, ' '); // excess white space
    const sqlStatements = splitStringByNotQuotedSemicolon(sqlReducedToStatements);

    for (const sql of sqlStatements) {
      await prisma.$executeRawUnsafe(sql);
    }
  } catch (e) {
    console.error(e);
    process.exit(1);
  } finally {
    await prisma.$disconnect();
  }
}

function splitStringByNotQuotedSemicolon(input: string): string[] {
  const result = [];

  let currentSplitIndex = 0;
  let isInString = false;
  for (let i = 0; i < input.length; i++) {
    if (input[i] === "'") {
      // toggle isInString
      isInString = !isInString;
    }
    if (input[i] === ';' && !isInString) {
      result.push(input.substring(currentSplitIndex, i + 1));
      currentSplitIndex = i + 2;
    }
  }

  return result;
}

void main();
```

**Comment by mon-jai** [comment]
@BjoernRave The following statement works for me. It works unless you have multiple SQL statements in the same line.

```js
const sqlStatements = rawSql
  .split("\n")
  .map(line => line.trim())
  .filter(line => line.length > 0 && !line.startsWith("--"))
  .join("\n")
  .replaceAll(/([^;])\n/g, (_match, p1) => p1)
  .split("\n")
```

**Comment by ludralph** [comment]
Hi @BjoernRave,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
